test/e2e/framework/suites.go

```diff
@@ -118,21 +118,29 @@ func SetupSuite() {
 		e2elog.Logf("kube-apiserver version: %s", serverVersion.GitVersion)
 	}
 
+	if TestContext.NodeKiller.Enabled {
+		nodeKiller := NewNodeKiller(TestContext.NodeKiller, c, TestContext.Provider)
+		go nodeKiller.Run(TestContext.NodeKiller.NodeKillerStopCh)
+	}
+}
+
+// SetupSuiteAllNodes is the boilerplate that can be used to setup ginkgo test suites, on the SynchronizedBeforeSuite step.
+// There are certain operations we only want to run once per overall test invocation on all nodes
+// such as making some global variables accessible to all parallel executions
+// Because of the way Ginkgo runs tests in parallel, we must use SynchronizedBeforeSuite
+func SetupSuiteAllNodes() {
 	// Obtain the default IP family of the cluster
 	// Some e2e test are designed to work on IPv4 only, this global variable
 	// allows to adapt those tests to work on both IPv4 and IPv6
-	// TODO(dual-stack): dual stack clusters should pass full e2e testing at least with the primary IP family
+	// TODO: dual-stack
 	// the dual stack clusters can be ipv4-ipv6 or ipv6-ipv4, order matters,
 	// and services use the primary IP family by default
-	// If we´ll need to provide additional context for dual-stack, we can detect it
-	// because pods have two addresses (one per family)
+	c, err := LoadClientset()
+	if err != nil {
+		klog.Fatal("Error loading client: ", err)
+	}
 	TestContext.IPFamily = getDefaultClusterIPFamily(c)
```

Comment:
it took me a while to discover this, it turns out that this is [how ginkgo run specs in parallel ](https://onsi.github.io/ginkgo/#parallel-specs)

> With SynchronizedBeforeSuite Ginkgo gives you a way to run some preliminary setup code on just one parallel node (Node 1) and other setup code on all nodes. Ginkgo synchronizes these functions and guarantees that node 1 will complete its preliminary setup before the other nodes run their setup code. Moreover, Ginkgo makes it possible for the preliminary setup code on the first node to pass information on to the setup code on the other nodes.

Previously ,TestContext.IPFamily was set on `framework.SetupSuite()`, but it turns out that part of the code only runs on just one parallel node in the first function of `SynchronizedBeforeSuite()`, thus the variable wasn't available for the rest of the nodes.

> SynchronizedBeforeSuite must be passed two functions. The first must return []byte and the second must accept []byte. When running with multiple nodes the first function is only run on node 1. When this function completes, all nodes (including node 1) proceed to run the second function and will receive the data returned by the first function. In this example, we use this data-passing mechanism to forward the database’s address (set up on node 1) to all nodes

I've decided to run the code to the get the variable on the second function from SynchronizedBeforeSuite() because maybe can be useful to other tests in the future, instead of passing the result using the []byte output

```
var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
	// Reference common test to make the import valid.		// Reference common test to make the import valid.
	commontest.CurrentSuite = commontest.E2E		commontest.CurrentSuite = commontest.E2E
	framework.SetupSuite()		framework.SetupSuite()
	return nil		return nil
}, func(data []byte) {	}, func(data []byte) {
	// Run on all Ginkgo nodes		// Run on all Ginkgo nodes
	framework.SetupSuiteAllNodes()
})	})
```

Reply:
> So `TestContext.IPFamily` is set _by_ the first parallel node, but does that mean it's not readable from other parallel nodes after the fact? `TestContext` a global var so I assumed it is.

right, actually I was thinking the same when I've implemented it but seems that `TestContext` is global per node executor, not global for all the nodes.

Reply:
The first node cannot modify `TestContext` in other nodes because each node runs in a separate process.

`AfterReadingAllFlags` is called by all processes, processing the same input, and thus the result is the same on all nodes. The `getDefaultClusterIPFamily(c)` call could work there, but I think it's better to call it where it is called now because of the need to connect to the API server. Doing that in `AfterReadingAllFlags` "feels" a bit too early.